[WIP] Update text classification template labels in DatasetInfo __post_init__

lewtun opened the pull request:
Hi @yjernite and @lhoestq, here's a first stab at the suggestion discussed in #2389 to update the `labels` of the `TextClassification` template in the `DatasetInfo.__post_init__`.

One problem I've spotted is that my current implementation introduces state into the `__post_init__`: 

* When we call `load_dataset`, `DatasetInfo.features` are the "raw" features without any casting so we can access the column names by the `label_column` specified in `TextClassification`
* When we call `Dataset.prepare_for_task` we run into a problem because the `DatasetInfo.features` are first cast into the new schema which triggers a `KeyError` when we update the infos [here](https://github.com/huggingface/datasets/blob/8b2a78520828e0cc13c14a31f413a5395ef25110/src/datasets/arrow_dataset.py#L1959).

What do you think? I did this a bit quickly, so maybe I'm overlooking something obvious :)

mariosasko commented:
If I'm not mistaken, one way to fix this would be to drop the task templates when copying the info by inserting `dataset.info.task_templates = None` before the `Dataset.cast` call in `Dataset.prepare_for_task`. Moreover, we should do this change independently of the KeyError being raised because currently the following is possible:
```python
dset = load_dataset("some_dataset") # let's say 'some_dataset' supports text classification and question answering
dset_tc = dset.prepare_for_task("text-classification")
dset_tc.preprare_for_task("question-answering") # this should raise an error because the schema is no longer valid for this task; currently this fails on 'rename_columns'
```
I see 2 options:
1. to drop the task templates after the first `Dataset.prepare_for_task` call
2. to save only the tasks compatible with the new schema after Dataset.prepare_for_task` (but then we have to update the column names of the compatible tasks to make sure the column mapping is still valid)

lewtun commented:
> If I'm not mistaken, one way to fix this would be to drop the task templates when copying the info by inserting `dataset.info.task_templates = None` before the `Dataset.cast` call in `Dataset.prepare_for_task`. Moreover, we should do this change independently of the KeyError being raised because currently the following is possible:
> 
> ```python
> dset = load_dataset("some_dataset") # let's say 'some_dataset' supports text classification and question answering
> dset_tc = dset.prepare_for_task("text-classification")
> dset_tc.preprare_for_task("question-answering") # this should raise an error because the schema is no longer valid for this task; currently this fails on 'rename_columns'
> ```
> 
> I see 2 options:
> 
> 1. to drop the task templates after the first `Dataset.prepare_for_task` call
> 2. to save only the tasks compatible with the new schema after Dataset.prepare_for_task` (but then we have to update the column names of the compatible tasks to make sure the column mapping is still valid)

thanks for the great idea @mariosasko and for spotting the problem with sequential task preparation! i am in favour of your option (1) since it is simple and saves us from having to keep track of the column mappings across multiple steps. 

i've implemented the change and refactored the tests to account for the new approach (including a new test that the templates are flushed after we call `prepare_for_task`). perhaps the slightly inelegant aspect here is that if we want to allow the user to set `labels` in the `TextClassification` template, then we have two places (`DatasetInfo.__post_init__` and `TextClassification.__post_init__`) where we need to update `label_schema`. 

on the other hand, dropping `labels` from the `TextClassification` signature would have the nice effect that users only have to think about column names when defining their tasks.

in any case, i think it would be a good idea to merge #2376 soon as the current PR is touching a lot of the same places in the codebase 😄

lewtun commented on datasets/emotion/emotion.py:
```diff
@@ -44,6 +45,12 @@ def _info(self):
             supervised_keys=("text", "label"),
             homepage=_URL,
             citation=_CITATION,
+            task_templates=[
```
this is here for demo purposes only and appears to be the reason why the tests are failing. i'll remove it if we decide to merge this PR

lewtun commented:
cc @SBrandeis who might also be interested in this feature :)

lhoestq commented:
Tests are failing only because the `emotion` dataset card doesn't pass our dataset card validator (tags are missing), you can ignore this since it's unrelated to this PR.

lhoestq commented on src/datasets/info.py:
```diff
@@ -166,6 +168,13 @@ def __post_init__(self):
+            for idx, template in enumerate(self.task_templates):
+                if isinstance(template, TextClassification) and self.features is not None:
+                    # Cast labels to tuple to enable hashing of task template
+                    object.__setattr__(template, "labels", tuple(sorted(self.features[template.label_column].names)))
```
Maybe raise an error if `template.labels` is not None and and doesn't match the labels in info.features ?

If users want to create a custom TaskTemplate with custom labels, then they must match the dataset labels.
e.g. this should fail:

```python
ds = load_dataset("emotion")
labels = ['label_that_doesnt_exist', 'another_one_that_doesnt_exist']
tc = TextClassification(text_column="text", label_column="label", labels=labels)
ds.prepare_for_task(tc)
```

lewtun replied:
> Maybe raise an error if `template.labels` is not None and and doesn't match the labels in info.features ?

Great idea! I decided to raise the error in `prepare_for_task` because `template.labels` is `None` by default in `DatasetInfo.__post_init__` and is the reason we set the attribute there with:

```python
# Cast labels to tuple to enable hashing of task template                   
 template.__dict__["labels"] = tuple(sorted(self.features[template.label_column].names))
```

I've also added a test to verify that we raise the error as expected (I also discovered some of my previous assertion tests were passing for the wrong reason 😅 ).

Perhaps one drawback from this approach is that we force the user to explicitly define the labels if they want to create their own `TextClassification` instance. On the other hand, being explicit is the Python way 😄 

> Also maybe we should create another instance of `TextClassification`, so that the user could reuse a TaskTemplate with unset labels later, instead of updating the current object.

I think that by raising the error in `prepare_for_task`, we no longer need a new instance because an instance with unset labels will raise a `ValueError` and not be modified. Please correct me if I misunderstood your remark.

lhoestq replied:
Thanks :) Let's allow labels to be None and sync with the features when appropriate (currently in DatasetInfo.__post_init__)

Regarding the new instance: I created a new instance in the changes I pushed.

Looks all good now

lhoestq reviewed (commented):
Looks great ! I added 2 comments:

lewtun commented on src/datasets/arrow_dataset.py:
```diff
@@ -1416,10 +1418,18 @@ def prepare_for_task(self, task: Union[str, TaskTemplate]) -> "Dataset":
         dataset = self.remove_columns(columns_to_drop)
         dataset = dataset.rename_columns(column_mapping)
+        # We found a template so now flush `DatasetInfo` to skip the template update in `DatasetInfo.__post_init__`
+        dataset.info.task_templates = None
```
It's actually worse than I anticipated: subsequent calls to `load_dataset` produce no templates

```
ds1 = load_dataset("./datasets/emotion/", split="train")
# task templates not empty (expected behaviour)
assert ds1.info.task_templates
ds2 = load_dataset("./datasets/emotion/", split="train")
# fails - task templates are empty
assert ds2.info.task_templates
```

lewtun replied:
It turns out that losing the templates on subsequent calls to `load_dataset` is not connected to `prepare_for_task`, but is actually due to the logic in `DatasetInfo.__post_init__`: I can reproduce the above example on `master` as well

Need to investigate the `__post_init__` first

lhoestq commented on src/datasets/tasks/text_classification.py:
```diff
@@ -23,30 +24,26 @@ def __get__(self, obj, objtype=None):
 
 @dataclass(frozen=True)
 class TextClassification(TaskTemplate):
-    task: ClassVar[str] = "text-classification"
+    # `task` is not a ClassVar since we want it to be part of the `asdict` output for JSON serialization
+    task: str = "text-classification"
```
This fixed your issue when reloading the dataset @lewtun 
Indeed `task_template_from_dict` wasn't able to reload the template from json data since the "task" field was missing

lewtun replied:
amazing, thanks for spotting and fixing this!

lhoestq commented on src/datasets/info.py:
```diff
@@ -167,6 +170,16 @@ def __post_init__(self):
 
+        # Insert labels and mappings for text classification
+        if self.task_templates is not None:
+            self.task_templates = list(self.task_templates)
```
I'm making a new list here to avoid side effects

lewtun commented:
@lhoestq @SBrandeis i've fixed the tests and think this is now in a good state for another review :)

lhoestq commented on tests/test_arrow_dataset.py:
```diff
@@ -702,8 +702,12 @@ def test_concatenate_with_no_task_templates(self, in_memory):
-        task_template = TextClassification(text_column="text", label_column="labels", labels=["pos", "neg"])
-        info = DatasetInfo(task_templates=task_template)
+        labels = ["neg", "pos"]
+        task_template = TextClassification(text_column="text", label_column="labels", labels=labels)
+        info = DatasetInfo(
+            features=Features({"text": Value("string"), "labels": ClassLabel(names=labels)}),
+            task_templates=task_template,
+        )
```
Maybe add a test to make sure that the DatasetInfo post init does pass the labels to the task templates, so that we can instantiate the task_template without labels as in the demo ?

lewtun replied:
good idea! i've now set up the unit test to cover the following cases:

1. `DatasetInfo` defined with a `TextClassification` template without `labels`
2. `DatasetInfo` defined with a `TextClassification` template with `labels` (this one might be overkill, but best to be safe)
3. Passing a `TextClassification` template with labels directly to `prepare_for_task` (the no labels case is covered in `test_task_with_incompatible_templates`)

lhoestq reviewed (approved):
Looks all good ! Thanks for the fix :)

lhoestq commented:
Maybe @SBrandeis you can also take a look to make sure you're fine with it ?

mariosasko commented on src/datasets/tasks/text_classification.py:
```diff
@@ -23,30 +24,26 @@ def __get__(self, obj, objtype=None):
+            assert len(self.labels) == len(set(self.labels)), "Labels must be unique"
+            # Cast labels to tuple to allow hashing
+            self.__dict__["labels"] = tuple(sorted(self.labels))
+            self.__dict__["label_schema"] = copy.deepcopy(self.label_schema)
```
Should be safe to remove the `FeaturesWithLazyClassLabel` descriptor after this to reduce the code complexity (and then the deepcopy call can be replaced with `self.label_schema.copy()`).

lewtun replied:
yes i agree, thanks for the suggestion! since we now create a new instance of `TextClassification` in `DatasetInfo.__post_init__` I think we should be safe from the side-effects you spotted earlier

mariosasko reviewed (commented):
Few suggestions.

lewtun commented on src/datasets/info.py:
```diff
@@ -167,6 +170,16 @@ def __post_init__(self):
+                    labels = self.features[template.label_column].names
+                    self.task_templates[idx] = TextClassification(
+                        text_column=template.text_column, label_column=template.label_column, labels=labels
+                    )
```
good idea!
